cpp/src/groupby/sort/group_merge_m2.cu

```diff
@@ -0,0 +1,187 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/dictionary/detail/iterator.cuh>
+#include <cudf/structs/structs_column_view.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/reduce.h>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+
+namespace {
+/**
+ * @brief Functor to accumulate (merge) all partial results corresponding to the same key into a
+ * final result storing in its member variables. It performs merging for the partial results of
+ * `COUNT_VALID`, `MEAN`, and `M2` at the same time.
+ */
+template <class ResultType>
+struct accumulate_fn {
+  size_type n_a;
+  ResultType mean_a;
+  ResultType M2_a;
+
+  void __device__ operator()(size_type const n_b,
+                             ResultType const mean_b,
+                             ResultType const M2_b) noexcept
+  {
+    if (n_b == 0) { return; }
+
+    auto const n_ab  = n_a + n_b;
+    auto const delta = mean_b - mean_a;
+    M2_a +=
+      M2_b + (delta * delta) * static_cast<ResultType>(n_a) * static_cast<ResultType>(n_b) / n_ab;
+    mean_a = (mean_a * n_a + mean_b * n_b) / n_ab;
+    n_a    = n_ab;
+  }
+};
+
+/**
+ * @brief Functor to merge partial results of `COUNT_VALID`, `MEAN`, and `M2` aggregations
+ * for a given group (key) index.
+ */
+template <class ResultType>
+struct merge_fn {
+  size_type const* const d_offsets;
+  size_type const* const d_counts;
+  ResultType const* const d_means;
+  ResultType const* const d_M2s;
+
+  auto __device__ operator()(size_type const group_idx) noexcept
+  {
+    auto const start_idx = d_offsets[group_idx], end_idx = d_offsets[group_idx + 1];
+
+    // This case should never happen, because all groups are non-empty due to the given input.
+    // Here just to make sure we cover this case.
+    if (start_idx == end_idx) {
+      return thrust::make_tuple(size_type{0}, ResultType{0}, ResultType{0}, int8_t{0});
+    }
+
+    // Firstly, this stores (valid_count, mean, M2) of the first partial result.
+    // Then, it accumulates (merges) the remaining partial results into it.
+    // Note that, if `n_a == 0` then `mean_a` and `M2_a` will be null.
+    // Thus, in such situations, we need to set zero for them before accumulating partial results.
+    auto const n_a    = d_counts[start_idx];
+    auto const mean_a = n_a > 0 ? d_means[start_idx] : ResultType{0};
+    auto const M2_a   = n_a > 0 ? d_M2s[start_idx] : ResultType{0};
```

Comment:
Since the logic here repeats below, maybe it should be moved into a helper function. One way to enable this would be to make a struct to hold the partial results:
```
template <class ResultType>
struct partial_result{
  size_type n;
  ResultType mean;
  ResultType M2;
};
```
And then the helper function can be implemented as
```
auto get_partial_result = [&](size_type index){
  auto const n    = d_counts[index];
  auto const mean = n > 0 ? d_means[index] : ResultType{0};
  auto const M2   = n > 0 ? d_M2s[index] : ResultType{0};
  return {n, mean, M2};};
```
With this, `accumulate_fn` can store the result as `partial_result a;` and its operator can be declared as `operator()(partial_result const& b);`